pkg/operators/uidgidresolver/usergroupcache.go

```diff
@@ -0,0 +1,261 @@
+		if err != nil {
+			log.Warnf("UserGroupCache: close watcher: %v", err)
+		}
+		// Wait until the loop is finished, should be fast
+		<-cache.loopFinished
+		cache.watcher = nil
+	}
+}
+
+func (cache *userGroupCache) Stop() {
+	cache.useCountMutex.Lock()
+	defer cache.useCountMutex.Unlock()
+
+	// We are the last user, stop everything
+	if cache.useCount == 1 {
+		cache.Close()
+	}
+	cache.useCount--
+}
+
+func (cache *userGroupCache) watchUserGroupLoop() {
+	defer close(cache.loopFinished)
+	for {
+		select {
+		case event, ok := <-cache.watcher.Events:
+			if !ok {
+				log.Warnf("UserGroupCache: watcher event not ok")
+				return
+			}
+			cache.handleEvent(event)
+		case err, ok := <-cache.watcher.Errors:
+			if !ok {
+				if err == nil {
+					// Watcher closed
+					return
+				}
+				log.Warnf("UserGroupCache: watcher error not ok: %v", err)
+				return
+			}
+			log.Warnf("UserGroupCache: watcher error: %v", err)
+		}
+	}
+}
+
+func (cache *userGroupCache) handleEvent(event fsnotify.Event) {
+	// Filter out chmod events first, to keep string comparisons to a minimum
+	if event.Has(fsnotify.Chmod) {
+		return
+	}
+
+	targetFilePath := ""
+	var resourceCache cachedmap.CachedMap[uint32, string]
+	if event.Name == fullPasswdPath {
+		targetFilePath = fullPasswdPath
+		resourceCache = cache.userCache
+	} else if event.Name == fullGroupPath {
+		targetFilePath = fullGroupPath
+		resourceCache = cache.groupCache
+	} else {
+		return
+	}
+
+	var targetFile *os.File
+	if event.Has(fsnotify.Remove) || event.Has(fsnotify.Rename) {
+		targetFile = nil
+	} else if event.Has(fsnotify.Write) || event.Has(fsnotify.Create) {
+		var err error
+		targetFile, err = os.OpenFile(targetFilePath, os.O_RDONLY, 0)
+		if err != nil {
+			log.Warnf("UserGroupCache: open target file: %v", err)
+			return
+		}
+		defer targetFile.Close()
+	} else {
+		// Ignore all other events
+		return
+	}
+
+	updateEntries(targetFile, resourceCache)
+}
+
+func updateEntries(file *os.File, resourceCache cachedmap.CachedMap[uint32, string]) {
+	oldEntries := make(map[uint32]struct{})
+	for _, id := range resourceCache.Keys() {
+		oldEntries[id] = struct{}{}
+	}
+
+	if file != nil {
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			line := strings.TrimLeft(scanner.Text(), " \t")
+			if len(line) == 0 || line[0] == '#' {
+				continue
+			}
+			split := strings.Split(line, ":")
+			// We are interested only in the first and third field
+			if len(split) < 3 {
+				continue
+			}
+			name := split[0]
+			id_u64, err := strconv.ParseUint(split[2], 10, 32)
+			if err != nil {
+				log.Warnf("UserGroupCache: convert id: %v", err)
+				continue
+			}
+			id := uint32(id_u64)
+			delete(oldEntries, id)
+			resourceCache.Add(id, &name)
+		}
+	}
+
+	for id := range oldEntries {
+		resourceCache.Remove(id)
+	}
+}
+
+func (cache *userGroupCache) GetUsername(uid uint32) string {
+	name := cache.userCache.Get(uid)
+	if name != nil {
+		return *name
+	}
```

Comment:
This seems weird. What about making the `cachedmap` return `T` instead of `*T` on the `Get()` (and other) methods, then it's up to the user whether to store pointers or values. In this case, we could store `string` directly, removing this logic here.

Reply:
TIL: I thought you can't instantiate a new object of a generic type parameter.

This won't compile. That is why I went with pointers
```
func (c *cachedMap[Key, T]) Get2(key Key) (T, error) {
    ....
    // Didn't find key
    return T{}
}
```

But there seems to be the following way, which compiles:
```
func (c *cachedMap[Key, T]) Get2(key Key) (T, error) {
    ....
    // Didn't find key
    var zeroValue T
    return zeroValue
}
```